fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -2491,6 +2523,57 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
         return bundle;
     }
 
+    /**
+     * For a chained search, check for existence of a versioned reference in the field associated with the
+     * chain search parameter for any of the specified resources.
+     *
+     * @param resourceType
+     *            The search resource type.
+     * @param queryParameter
+     *            The chained query parameter.
+     * @param resources
+     *            The list of resources to check for versioned references.
+     * @return
+     *            A list of Issues, one per resource in which a versioned reference is found.
+     * @throws Exception
+     */
+    private List<Issue> checkForVersionedReference(Class<? extends Resource> resourceType, QueryParameter queryParameter, List<Resource> resources) throws Exception {
+        List<Issue> issues = new ArrayList<>();
+
+        // Since the chained search was successful, we can assume the search parameter exists and is valid and is of type Reference
+        SearchParameter chainSearchParm = null;
+        if (!Resource.class.equals(resourceType)) {
+            chainSearchParm = SearchUtil.getSearchParameter(resourceType, queryParameter.getCode());
+        }
+        for (Resource resource : resources) {
+            FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
+            EvaluationContext evaluationContext = new EvaluationContext(resource);
+            SearchParameter searchParameter = null;
+            try {
+                if (chainSearchParm != null) {
+                    searchParameter = chainSearchParm;
+                } else {
+                    searchParameter = SearchUtil.getSearchParameter(resource.getClass(), queryParameter.getCode());
+                }
+                Collection<FHIRPathNode> nodes = evaluator.evaluate(evaluationContext, searchParameter.getExpression().getValue());
+                for (FHIRPathNode node : nodes) {
+                    Reference reference = node.asElementNode().element().as(Reference.class);
+                    ReferenceValue rv = ReferenceUtil.createReferenceValueFrom(reference, ReferenceUtil.getBaseUrl(null));
+                    if (rv.getVersion() != null &&
+                            (rv.getTargetResourceType() == null || rv.getTargetResourceType().equals(queryParameter.getModifierResourceTypeName()))) {
+                        String msg = "Resource with id '" + resource.getId() + "' contains a versioned reference in an element used for chained search, but chained search does not act on versioned references.";
```

Comment:
Maybe include the resource type in this message as well?  Or will it be always be easy to discern from the request?

Reply:
I guess since its only for forward-chaining, it should always be the resource type that is being searched.  So probably no need to specify it here.